Jitsi custom OSGi classloader issues: fix way bundles are loaded in launcher

cobratbq opened the pull request:
I seem to run into classloader-related issues if I build and run the Jitsi code-base as-is.
I don't know if that's an error of how I run it, but this simple fix solves significant issues for me. (@ibauersachs do you know?)

There's some other minor changes.

ibauersachs commented:
What's the issue you encounter?

The mainWithCl stuff is necessary so that native bundles are properly loaded IIRC.

cobratbq commented:
> What's the issue you encounter?

1. spellchecker bundle fails to load successfully, prevents GUI from loading because the spellchecker UI component fails.
2. If I create an XMPP account, then connect it, it fails to load due to an issue with StaxXMLParser. (See stack trace below.)

My theory was (eventually) that this might be a conflict of classloaders, because the type is a subtype. Given that (1.) also occurred. I already tried things like explicitly adding dependency on the Xml-artifact (without '-stax' suffix). I looked into OSGi. Found your advice to someone about using Spifly, but didn't help. Then found the custom loading in the launcher.

```
2023-05-28 23:27:39.703 [DEBUG] [ForkJoinPool.commonPool-worker-3] net.java.sip.communicator.impl.protocol.jabber.JabberActivator - Got service UIServiceImpl, starting now
2023-05-28 23:27:39.703 [INFO ] [ForkJoinPool.commonPool-worker-1] net.java.sip.communicator.plugin.advancedconfig.AdvancedConfigurationPanel - Handling registration of a new Configuration Form.
2023-05-28 23:27:39.752 [ERROR] [ForkJoinPool.commonPool-worker-3] org.jitsi.impl.osgi.framework.launch.EventDispatcher - Error dispatching event
java.util.ServiceConfigurationError: org.jivesoftware.smack.xml.XmlPullParserFactory: org.jivesoftware.smack.xml.stax.StaxXmlPullParserFactory not a subtype
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at org.jivesoftware.smack.xml.SmackXmlParser.getXmlPullParserFactory(SmackXmlParser.java:40)
	at org.jivesoftware.smack.xml.SmackXmlParser.newXmlParser(SmackXmlParser.java:65)
	at org.jivesoftware.smack.util.PacketParserUtils.getParserFor(PacketParserUtils.java:80)
	at org.jivesoftware.smack.SmackInitialization.processConfigFile(SmackInitialization.java:159)
	at org.jivesoftware.smack.SmackInitialization.processConfigFile(SmackInitialization.java:154)
	at org.jivesoftware.smack.SmackInitialization.<clinit>(SmackInitialization.java:103)
	at org.jivesoftware.smack.Smack.getVersion(Smack.java:38)
	at org.jivesoftware.smack.Smack.ensureInitialized(Smack.java:64)
	at org.jivesoftware.smack.provider.ProviderManager.<clinit>(ProviderManager.java:125)
	at org.jitsi.xmpp.extensions.jingle.JingleIQProvider.<init>(JingleIQProvider.java:48)
	at net.java.sip.communicator.impl.protocol.jabber.ProviderManagerExt.load(ProviderManagerExt.java:50)
	at net.java.sip.communicator.impl.protocol.jabber.ProtocolProviderFactoryJabberImpl.<init>(ProtocolProviderFactoryJabberImpl.java:54)
	at net.java.sip.communicator.impl.protocol.jabber.JabberActivator.startWithServices(JabberActivator.java:174)
	at net.java.sip.communicator.util.osgi.DependentActivator.addingService(DependentActivator.java:103)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:871)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:903)
	at org.jitsi.impl.osgi.framework.launch.EventDispatcher$Command.run(EventDispatcher.java:121)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

> The mainWithCl stuff is necessary so that native bundles are properly loaded IIRC.

Just to check, did you notice that the logic in `mainWithCl` is still present, just moved to `main`? I just removed the _invoking-with-bundle-classloader_ logic. From what I can tell, when the framework gets initialized, it already takes into account that it gets passed in a plain classloader and seems to operate fine regardless.

To clarify, to run I do: `mvn clean verify -DskipTests; java -cp 'lib/:target/bundles/*' net.java.sip.communicator.launcher.Jitsi` and that's when the above happens.

cobratbq commented:
Converted to draft: I made a bad assumption on part of the spellchecker, i.e. it's still broken. I want to double-check everything first.

ibauersachs commented:
The launcher classpath was missing the entire xmpp-package. Doh.
I've pushed some fixes and I could connect to a local OpenFire server with them. But I didn't try the spellchecking yet. Not surprised if there are errors left from the conversion of real OSGi to the simpler/fake jitsi-osgi.

The changes you made in the main class don't work: the point of the reflective invocation of mainWithCl is that the actual startup is already made from an OSGi classloader. Resource loading , the launch arguments, and other stuff will otherwise break.

cobratbq commented:
> The launcher classpath was missing the entire xmpp-package. Doh. I've pushed some fixes and I could connect to a local OpenFire server with them. But I didn't try the spellchecking yet. Not surprised if there are errors left from the conversion of real OSGi to the simpler/fake jitsi-osgi.

Sure, it makes things a bit more fragile, but at the same time the code does not demand very exotic features.

It is curious: you say that you need to add the XMPP packages to the launcher dependencies? That shouldn't be necessary, as you would expect one of Jitsi's XMPP packages to load anything necessary as dependent package. (I.e. not directly a jitsi feature, therefore need not be available.)

My line of reasoning was that the framework internally already tackles the necessary parts for "installing bundles", etc., therefore wrapping with `BundleClassLoader` is unnecessary.

> The changes you made in the main class don't work: the point of the reflective invocation of mainWithCl is that the actual startup is already made from an OSGi classloader. Resource loading , the launch arguments, and other stuff will otherwise break.

I am not surprised to hear this. Still, for me, the classloading issues that I had previously, disappeared. I understand that it does not start with the `BundleClassLoader` as a root classloader. That's why the manual, reflective invocation was necessary. Using actual classloader + specifying same classloader in reflections was enough for me.

It's curious. I'm going to have another look.

ibauersachs commented:
It's not Smack that was missing, but the entire protocol-jabber project. This is a fully optional part, just like SIP or IRC. You probably didn't notice this as the project was built nonetheless and adding bundles/* to the classpath in the script actually made it available. But it wasn't as part of the Maven classpath of the launcher submodule.

Bundle loading is indeed done within the 'framework', but the LaunchArgs would then be in a different classloader than the bundles that need it. Plus the other stuff with resources, native libs, etc. I can't remember all the details as it's been a while, but I know it solved more problems than it created.

Actually, I'd like to get rid of all OSGi stuff that's beyond the service registry. But that's probably something I won't touch that, it's not worthy enough of my time anymore.

cobratbq commented:
> > It's not Smack that was missing, but the entire protocol-jabber project. This is a fully optional part, just like SIP or IRC. You probably didn't notice this as the project was built nonetheless and adding bundles/* to the classpath in the script actually made it available. But it wasn't as part of the Maven classpath of the launcher submodule.
> 
> Ah, right! That also makes sense in my understanding of it. :-P

Right, checked.

> > Bundle loading is indeed done within the 'framework', but the LaunchArgs would then be in a different classloader than the bundles that need it. Plus the other stuff with resources, native libs, etc. I can't remember all the details as it's been a while, but I know it solved more problems than it created.
> 
> I prly will have a look at this, if only to see if I understand how it works.

The `FrameworkImpl` functions as the "root bundle". When the classloader is included in a bundle-to-be-installed, it is always wrapped in a `BundleClassLoader`. I have adapted the `reflections` function-calls to include the same classloader. (My intuition/educated guess is that the `BundleClassLoader` exists primarily to compartmentalize the loaded classes, such that service-loading actions are possible. Then the root classloader - part of framework - does not strictly need wrapped.)

cobratbq commented:
Spellchecker fixed. (See 486ebaa33759cbf48adbd7af96c45dfee3b19fde.)

cobratbq commented:
Let me know if you guys are even interested in any improvements. I fully understand if focus is on other things. I just want to avoid others spending the same time fixing the same issues.

ibauersachs commented on modules/launcher/pom.xml:
```diff
@@ -275,6 +286,11 @@
       <artifactId>jitsi-protocol-irc</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
```
This is already in master.

ibauersachs commented on modules/plugin/otr/src/main/java/net/java/sip/communicator/plugin/otr/authdialog/KnownFingerprintsTableModel.java:
```diff
@@ -75,12 +75,12 @@ public KnownFingerprintsTableModel()
 
         // Populate contacts.
-        for (int i = 0; i < protocolProviderRefs.length; i++)
+        for (final ServiceReference<?> protocolProviderRef : protocolProviderRefs)
```
Start using `var` where possible if you touch things anyway

cobratbq commented on run-build.sh:
This was a cheap workaround. Can you fill me in on how I am supposed to build and run the desktop app? I assume there is an expected way?!

I see predominantly maven, but also some gradle scripts. `maven-exec` is a reasonable option, but the run-script represents "whatever-worked-for-me".

ibauersachs replied:
To build: `mvn package`.
To run: use e.g. the stored IntelliJ launch configuration.

I realize there's tons of preferences how people setup their dev-environment so you might not be using IntelliJ. If you want to keep the script, please do it in your personal working directory, and e.g. add it to .git/.gitignore so it can stay there without being noticed by anybody else.

cobratbq commented on modules/plugin/spellcheck/src/main/java/net/java/sip/communicator/plugin/spellcheck/Parameters.java:
```diff
@@ -51,14 +51,13 @@ class Parameters
-                    RESOURCE_LOC);
-
-            InputStream stream = url.openStream();
+            final InputStream stream = Parameters.class.getClassLoader()
```
This case is unique and unnecessary use: it goes back to the bundle and then stays within the bundle. I agree that it would be nicer if a total OSGi implementation was available, but how often do you load resources across bundle boundaries? I suspect it is rare, and this case was false, so not even an occurrence.

cobratbq replied:
@ibauersachs to add, I hadn't dismissed your remark about fixing the custom OSGi. I am worried that the lack of interest in the desktop app may not warrant investing that much time. (That is the impression I am getting.) You mentioned before that we might simply move away from the more elaborate uses of OSGi and stick with plugin management.

I don't know enough about jitsi's user base or its OSGi use, so feel free to correct me. Is this more than a gut feeling for you? Like, do you have a list of (problematic) uses? Or something else actionable?

ibauersachs replied:
Well, any time put into Jitsi is probably just for fun or to learn something new by now.

No, I don't. The reason I moved away from full OSGi was that too many libraries don't have an OSGi manifest and all the problems that come with it (like rebundling). Also, it's way faster now. The drawback is of course that Jitsi was relying on real OSGi and stuff might silently be broken, like the spellchecker you found.

ibauersachs commented:
> Let me know if you guys are even interested in any improvements. I fully understand if focus is on other things. I just want to avoid others spending the same time fixing the same issues.

Anything that gets stuff working again is welcome. Preferably in small chunks as I'm currently not working a lot on it and if I can click merge on GitHub mobile it's off everybody's mind ;-)

ibauersachs commented on modules/launcher/src/main/java/net/java/sip/communicator/launcher/Jitsi.java:
```diff
@@ -101,17 +95,7 @@ public class Jitsi
      * @param args command line args if any
      * @throws Exception whenever it makes sense.
      */
-    public static void main(String[] args)
```
Keep this. I've checked again, and it really is needed, even with the other changes below. Set a breakpoint in any Activator loaded from the "framework", you'll see that it's loaded from AppClassLoader instead of the BundleClassLoader.

Why is that? Contrary to a real OSGi framework, this fake-thing doesn't isolate jars and their classes. Especially not the ones directly specified at the JVM level. Since _all_ of the classes are specified there, all classloading bubbles up (which is specified behavior of ClassLoaders). This breaks native library loading. You mentioned it worked for you, but this is not reliable, e.g. on Windows WASAPI still loads because JNA magically finds it at the right place, but other stuff is missing (jnffmpeg, hid, portaudio, ...).

cobratbq replied:
I investigated for myself. See comment.

Neustradamus commented:
Why not to merge directly this PR?

cobratbq commented:
The review comments were valid. I have not made time to fix those. I plan to pick this up but have been delaying. I will (probably) make a new PR when the changes have been made.

On July 19, 2023 5:46:54 PM UTC, Neustradamus ***@***.***> wrote:
>Why not to merge directly this PR?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/jitsi/jitsi/pull/801#issuecomment-1642501996
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

cobratbq commented:
@ibauersachs I investigated myself and have these findings. I cannot confirm your claim that the AppClassLoader is used directly. These results correspond with my earlier claim. See steps below.

1.  set breakpoint in `OtrActivator`.
2.  when on breakpoint, move up the stack to `FrameworkStartLevelImpl$Command`. (See first code-snippet below.)
3.  in memory-outline to the right for selected stack frame:
    - expand `this`,
    - expand `this$0`,
    - expand `bundles`,
    - expand `0` + `1` + `1/bundleActivators` + `1/classLoader` + `2` + `2/bundleActivators` + `2/classLoader`
4.  this should roughly represent the second code snippet.
5.  observe:
    - First entry (i.e. `0`) is the Framework, with the classloader `{ClassLoaders$AppClassLoader@3790}`, as expected.
    - Second entry is bundle, in this case "net.java.sip.communicator.launcher.ShutdownTimeout", "net.java.sip.communicator.argdelegation.ArgDelegationActivator".
    - Second entry classloader is '{BundleClassLoader@4996}' with parent '{ClassLoaders$AppClassLoader@3790}`, i.e. the framework classloader, as expected.
    - Same for third entry (second bundle) and fourth entry (third bundle), etc.

```
<init>:124, OtrActivator (net.java.sip.communicator.plugin.otr)
newInstance0:-1, NativeConstructorAccessorImpl (jdk.internal.reflect)
newInstance:77, NativeConstructorAccessorImpl (jdk.internal.reflect)
newInstance:45, DelegatingConstructorAccessorImpl (jdk.internal.reflect)
newInstanceWithCaller:499, Constructor (java.lang.reflect)
newInstance:480, Constructor (java.lang.reflect)
start:467, BundleImpl (org.jitsi.impl.osgi.framework)
startLevelChanged:493, FrameworkImpl (org.jitsi.impl.osgi.framework.launch)
run:133, FrameworkStartLevelImpl$Command (org.jitsi.impl.osgi.framework.startlevel)
exec:1395, ForkJoinTask$RunnableExecuteAction (java.util.concurrent)
doExec$$$capture:373, ForkJoinTask (java.util.concurrent)
doExec:-1, ForkJoinTask (java.util.concurrent)
topLevelExec:1182, ForkJoinPool$WorkQueue (java.util.concurrent)
scan:1655, ForkJoinPool (java.util.concurrent)
runWorker:1622, ForkJoinPool (java.util.concurrent)
run:165, ForkJoinWorkerThread (java.util.concurrent)
```

```
this = {FrameworkStartLevelImpl$Command@3839} 
 listeners = {FrameworkListener[1]@3837} 
 startLevel = 3
 this$0 = {FrameworkStartLevelImpl@3814} 
  framework = {FrameworkImpl@3820} 
   logger = {Logger@3824} 
   bundles = {LinkedList@3825}  size = 86
    0 = {FrameworkImpl@3820} 
     logger = {Logger@3824} 
     bundles = {LinkedList@3825}  size = 86
     configuration = {ImmutableCollections$Map1@3826}  size = 1
     eventDispatcher = {EventDispatcher@3827} 
     frameworkStartLevel = {FrameworkStartLevelImpl@3814} 
     nextBundleId = 86
     nextServiceId = 25
     serviceRegistrations = {LinkedList@3828}  size = 24
     stopEvent = {Object@3829} 
     frameworkListeners = {Collections$SynchronizedRandomAccessList@3830}  size = 0
     bundleActivatorClasses = {ArrayList@3831}  size = 0
     bundleActivators = {ArrayList@3832}  size = 0
     headers = {Hashtable@3833}  size = 1
     bundleContext = {BundleContextImpl@3834} 
     bundleId = 0
     bundleStartLevel = {BundleStartLevelImpl@3835} 
     framework = null
     location = "file:/home/dev-otr/.m2/repository/org/jitsi/jitsi-android-osgi/2.0-3-gd59644a/jitsi-android-osgi-2.0-3-gd59644a.jar"
     classLoader = {ClassLoaders$AppClassLoader@3790} 
     state = 8
    1 = {BundleImpl@4906} 
     bundleActivatorClasses = {ArrayList@4990}  size = 2
      0 = {Class@2331} "class net.java.sip.communicator.launcher.ShutdownTimeout"
      1 = {Class@2334} "class net.java.sip.communicator.argdelegation.ArgDelegationActivator"
     bundleActivators = {ArrayList@4991}  size = 2
     headers = {Hashtable@4992}  size = 1
     bundleContext = {BundleContextImpl@4993} 
     bundleId = 1
     bundleStartLevel = {BundleStartLevelImpl@4994} 
     framework = {FrameworkImpl@3820} 
     location = "file:/home/dev-otr/dev/jitsi/modules/launcher/target/classes/"
     classLoader = {BundleClassLoader@4996} 
      parent = {ClassLoaders$AppClassLoader@3790} 
      nativeLibraries = {HashMap@4997}  size = 0
      nativeLibTempDir = null
      ucp = {URLClassPath@4998} 
      acc = {AccessControlContext@4999} 
      closeables = {WeakHashMap@5000}  size = 0
      pdcache = {ConcurrentHashMap@5001}  size = 0
      ClassLoader.parent = null
      name = null
      unnamedModule = {Module@5002} "unnamed module @145aa870"
      nameAndId = "org.jitsi.impl.osgi.framework.BundleClassLoader @15d49048"
      parallelLockMap = null
      package2certs = {ConcurrentHashMap@5004}  size = 0
      classes = {ArrayList@5005}  size = 0
      defaultDomain = {ProtectionDomain@5006} "ProtectionDomain  (null <no signer certificates>)\n org.jitsi.impl.osgi.framework.BundleClassLoader@15d49048\n <no principals>\n java.security.Permissions@686fd5f0 (\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.beans.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.apple.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.java.swing.plaf.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.beans")\n ("java.util.PropertyPermission" "java.specification.version" "read")\n ("java.util.PropertyPermission" "java.vm.vendor" "read")\n ("java.util.PropertyPermission" "path.separator" "read")\n ("java.util.PropertyPermission" "os.version" "read")\n ("java.util.PropertyPermission" "java.vendor.url" "read")\n ("java.util.PropertyPermission" "java.vm.name" "read")\n ("java.util.PropertyPermission" "java.vm.specification.version" "read")\n ("java.util.PropertyPermission" "os.name" "read")\n ("java.util.PropertyPermission" "java.version" "read")\n ("java.ut"
      packages = {ConcurrentHashMap@5007}  size = 0
      libraries = {NativeLibraries@5008} 
      assertionLock = {BundleClassLoader@4996} 
      defaultAssertionStatus = false
      packageAssertionStatus = null
      classAssertionStatus = null
      classLoaderValueMap = null
     state = 32
    2 = {BundleImpl@4907} 
     bundleActivatorClasses = {ArrayList@5014}  size = 1
     bundleActivators = {ArrayList@5015}  size = 1
      0 = {NimbuzzAvatarsActivator@5022} 
     headers = {Hashtable@5016}  size = 1
     bundleContext = {BundleContextImpl@5017} 
     bundleId = 2
     bundleStartLevel = {BundleStartLevelImpl@5018} 
     framework = {FrameworkImpl@3820} 
     location = "file:/home/dev-otr/dev/jitsi/modules/plugin/nimbuzzavatars/target/classes/"
     classLoader = {BundleClassLoader@5020} 
      parent = {ClassLoaders$AppClassLoader@3790} 
      nativeLibraries = {HashMap@5023}  size = 0
      nativeLibTempDir = null
      ucp = {URLClassPath@5024} 
      acc = {AccessControlContext@5025} 
      closeables = {WeakHashMap@5026}  size = 0
      pdcache = {ConcurrentHashMap@5027}  size = 0
      ClassLoader.parent = null
      name = null
      unnamedModule = {Module@5028} "unnamed module @40dd4c21"
      nameAndId = "org.jitsi.impl.osgi.framework.BundleClassLoader @a8ef162"
      parallelLockMap = null
      package2certs = {ConcurrentHashMap@5030}  size = 0
      classes = {ArrayList@5031}  size = 0
      defaultDomain = {ProtectionDomain@5032} "ProtectionDomain  (null <no signer certificates>)\n org.jitsi.impl.osgi.framework.BundleClassLoader@a8ef162\n <no principals>\n java.security.Permissions@28b3ea15 (\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.beans.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.apple.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.java.swing.plaf.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.beans")\n ("java.util.PropertyPermission" "java.specification.version" "read")\n ("java.util.PropertyPermission" "java.vm.vendor" "read")\n ("java.util.PropertyPermission" "path.separator" "read")\n ("java.util.PropertyPermission" "os.version" "read")\n ("java.util.PropertyPermission" "java.vendor.url" "read")\n ("java.util.PropertyPermission" "java.vm.name" "read")\n ("java.util.PropertyPermission" "java.vm.specification.version" "read")\n ("java.util.PropertyPermission" "os.name" "read")\n ("java.util.PropertyPermission" "java.version" "read")\n ("java.uti"
      packages = {ConcurrentHashMap@5033}  size = 0
      libraries = {NativeLibraries@5034} 
      assertionLock = {BundleClassLoader@5020} 
      defaultAssertionStatus = false
      packageAssertionStatus = null
      classAssertionStatus = null
      classLoaderValueMap = null
     state = 32
    3 = {BundleImpl@4908} 
     bundleActivatorClasses = {ArrayList@5037}  size = 1
     bundleActivators = {ArrayList@5038}  size = 1
      0 = {HIDActivator@5045} 
     headers = {Hashtable@5039}  size = 1
     bundleContext = {BundleContextImpl@5040} 
     bundleId = 3
     bundleStartLevel = {BundleStartLevelImpl@5041} 
     framework = {FrameworkImpl@3820} 
     location = "file:/home/dev-otr/dev/jitsi/modules/impl/hid/target/classes/"
     classLoader = {BundleClassLoader@5043} 
      parent = {ClassLoaders$AppClassLoader@3790} 
      nativeLibraries = {HashMap@5046}  size = 0
      nativeLibTempDir = null
      ucp = {URLClassPath@5047} 
      acc = {AccessControlContext@5048} 
      closeables = {WeakHashMap@5049}  size = 0
      pdcache = {ConcurrentHashMap@5050}  size = 0
      ClassLoader.parent = null
      name = null
      unnamedModule = {Module@5051} "unnamed module @3ee56aa6"
      nameAndId = "org.jitsi.impl.osgi.framework.BundleClassLoader @2eea88a1"
      parallelLockMap = null
      package2certs = {ConcurrentHashMap@5053}  size = 0
      classes = {ArrayList@5054}  size = 0
      defaultDomain = {ProtectionDomain@5055} "ProtectionDomain  (null <no signer certificates>)\n org.jitsi.impl.osgi.framework.BundleClassLoader@2eea88a1\n <no principals>\n java.security.Permissions@7f3d5c22 (\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.beans.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.apple.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.java.swing.plaf.*")\n ("java.lang.RuntimePermission" "accessClassInPackage.com.sun.beans")\n ("java.util.PropertyPermission" "java.specification.version" "read")\n ("java.util.PropertyPermission" "java.vm.vendor" "read")\n ("java.util.PropertyPermission" "path.separator" "read")\n ("java.util.PropertyPermission" "os.version" "read")\n ("java.util.PropertyPermission" "java.vendor.url" "read")\n ("java.util.PropertyPermission" "java.vm.name" "read")\n ("java.util.PropertyPermission" "java.vm.specification.version" "read")\n ("java.util.PropertyPermission" "os.name" "read")\n ("java.util.PropertyPermission" "java.version" "read")\n ("java.ut"
      packages = {ConcurrentHashMap@5056}  size = 0
      libraries = {NativeLibraries@5057} 
      assertionLock = {BundleClassLoader@5043} 
      defaultAssertionStatus = false
      packageAssertionStatus = null
      classAssertionStatus = null
      classLoaderValueMap = null
     state = 32
[..]
```

ibauersachs commented:
I don't know what exactly you tried, but it doesn't work. What you show in the stacktrace and inspection is a field on `BundleImpl`, which doesn't represent the actually used classloader. If you expand `bundleActivatorClasses/0/classLoader` you'll see the actual classloader of the activators.

Or easier, while still in `OtrActivator`, just inspect `getClass().getClassLoader()`.

**Via Reflection:**

![main-reflection](https://github.com/jitsi/jitsi/assets/949783/d6448a02-15fe-43cf-b9e3-0deb2da8d291)

**Your suggested change:**

![main-direct](https://github.com/jitsi/jitsi/assets/949783/c260555b-6206-4782-a517-fd7c4ac84ae2)

cobratbq commented:
> I don't know what exactly you tried, but it doesn't work. What you show in the stacktrace and inspection is a field on `BundleImpl`, which doesn't represent the actually used classloader. If you expand `bundleActivatorClasses/0/classLoader` you'll see the actual classloader of the activators.

Confirmed.

> Or easier, while still in `OtrActivator`, just inspect `getClass().getClassLoader()`. Edit: for the record, this is on JDK 11.0.20 (Temurin).

JDK versions shouldn't matter. Regardless, same outcome for me.

---

Okay, I am not sure we have the same goal in mind. And I think the discussion essentially becomes: "how is OSGi supposed to work?"

Your suggested change would result in _all_ bundles being loaded in the same BundleClassLoader instance, as well as the application. I think your goal is to have all bundles in separate BundleClassLoaders, essentially what `installBundle` does, but also for all the classes/types. However, that would mean you cut off access to the types from the Jitsi framework/application itself.

My suggestion loads the types of the direct dependencies, i.e. what Jitsi works with, in the AppClassLoader, and provides a `BundleContext` with isolated classloader for dependencies to be installed in. All services offered by the `BundleContext` _should_ use their own classloader.

At this point, my question is whether the suggested wrapping of `AppClassLoader` with a single `BundleClassLoader` makes a different (i.e. is or is not redundant), as:

1. your suggestion would still have a combined: Jitsi Framework + all bundles types.
1. how much class isolation do you want to achieve?

Additionally, the issue might start earlier:

We currently use `reflections` to search for the classes, then we figure out the file path, and then we "pretend" to install it anew. However, the classloader has already touched it at this point. (Hence Jitsi + all bundles in single BundleClassLoader.)

If this is the issue, we should search by filename and install that way, such that the classloader doesn't touch/initialize.

Right now, I run into a combination of limitations to understanding OSGi and confusing statements/results from what is happening in the Jitsi code: either the current code does not achieve the result you have in mind, or both suggestions achieve same nett outcome.

ibauersachs commented:
The goal is to get rid of OSGi as much as possible. However, since the whole application very deepöy relies on it, it's next to impossible, at least for the service registry.
Why?
Many dependencies do not have the required manifest, are broken or become randomly broken in an update. Working with native Java mechanism lile ServiceLoader is a pain.  IDE support is minimal. Setup is a pain.

The reason for all the foo with the filenames is because it's currently required to launch Jitsi and not throwing NPEs.

The classloader-per-bundle isolation is not required in Jitsi, if I wanted that, Felix would still be the framework not the current fake.

The only reason BundleClassLoader exists at all is for the native libraries as explained in a previous comment.
That could go away too if you go over all jars of the classpath, extract all natives like the BundleClassLoader would into a directory that is on java.library.path (that could even be done at build time).

What's _your_ goal?

cobratbq commented:
> What's _your_ goal?

Fix the code-base such that we can reliably build jitsi without needing to investigate and/or make changes to the code. Possibly add some error handling to prevent whole application from crashing/hanging/... when something is missing. (Such as in my case the issue with spellchecker.) I wanted to upstream the changes that I needed to make, and in the proper way, hence the discussion right now. Such that when someone checks out the codebase next, they will be able to get started immediately.

I think we were aligned on everything except the reflection/OSGi loading stuff. And, because it seems like we're contradicting each-other, I want(ed) to figure out how/what/why.

It's also why I checked whether or not the contributions are still wanted. (Not trying to revisit this question, just stating motivation.)

ibauersachs commented:
> > What's _your_ goal?
> 
> Fix the code-base such that we can reliably build jitsi without needing to investigate and/or make changes to the code. Possibly add some error handling to prevent whole application from crashing/hanging/... when something is missing. (Such as in my case the issue with spellchecker.) I wanted to upstream the changes that I needed to make, and in the proper way, hence the discussion right now. Such that when someone checks out the codebase next, they will be able to get started immediately.

That's precisely why I got rid of the real OSGi framework, Felix. E.g. no more manual Import-Package/Export-Package fiddling. No more repackaging of non-OSGified libs from Maven.

> I think we were aligned on everything except the reflection/OSGi loading stuff. And, because it seems like we're contradicting each-other, I want(ed) to figure out how/what/why.

The reflection stuff in `Main`, ugly as it is, makes everything much simpler.

> It's also why I checked whether or not the contributions are still wanted. (Not trying to revisit this question, just stating motivation.)

They are. But hopefully it's obvious that I request on stuff that doesn't work.

cobratbq commented:
Aha! Now I get what you're talking about. :-P The library-loading has been sprinkled with magic fairy dust. `BundleClassLoader#findLibrary` makes native libraries available/accessible upon searching for them.

So, basically, my understanding of OSGi is correct, but the code relies on an additional benefit that we rely on.

```OSUtils.java
if (OSUtils.class.getClassLoader().getClass().getName().contains("Bundle")) {
    loadLibrary = (libname, clazz) -> {
        System.loadLibrary(libname);
    };
} else {
    loadLibrary = JNIUtils::loadLibrary;
}
```

ibauersachs commented:
> Aha! Now I get what you're talking about. :-P The library-loading has been sprinkled with magic fairy dust. 

That magic fairy dust is what a OSGi framework normally does: https://github.com/apache/felix-dev/blob/e75de7154a69f03bf87a5024ddca414da1fcc8c9/framework/src/main/java/org/apache/felix/framework/BundleWiringImpl.java#L2481

> `BundleClassLoader#findLibrary` makes native libraries available/accessible upon searching for them. Most of the native libraries are loaded in the static initializer, so we cannot benefit from the bundle initialization itself.

The static initializer has nothing to do with bundle initialization. The only way to catch `System.loadLibrary` and provide a custom location after JVM startup (without reflection) is a custom classloader. If you know of anything else: awesome!

> So, basically, my understanding of OSGi is correct, but the code relies on the adapted native-library-loading provided in the `BundleClassLoader`. (I know that's what you said, but not with sufficient detail for it to make sense to me in terms of how we are restricted.)

I probably assumed you still know all that stuff from `back in the days™`...

cobratbq commented:
> > Aha! Now I get what you're talking about. :-P The library-loading has been sprinkled with magic fairy dust.
> 
> That magic fairy dust is what a OSGi framework normally does: https://github.com/apache/felix-dev/blob/e75de7154a69f03bf87a5024ddca414da1fcc8c9/framework/src/main/java/org/apache/felix/framework/BundleWiringImpl.java#L2481

Sure, except it was not obvious how to reconcile the two notions of "_getting rid of the full-blown OSGi runtime_" and "_we need to do this manual thing to get the OSGi to work correctly_". They seemed contradictory, as the classloader makes a significant difference while the OSGi framework was not bothered in the least.

> > `BundleClassLoader#findLibrary` makes native libraries available/accessible upon searching for them. Most of the native libraries are loaded in the static initializer, so we cannot benefit from the bundle initialization itself.
> 
> The static initializer has nothing to do with bundle initialization. The only way to catch `System.loadLibrary` and provide a custom location after JVM startup (without reflection) is a custom classloader. If you know of anything else: awesome!

The Java run-time leaves almost no wiggle-room. This also involves `libjitsi` which is shared among products IIUC, so I am not going to go there.

> > So, basically, my understanding of OSGi is correct, but the code relies on the adapted native-library-loading provided in the `BundleClassLoader`. (I know that's what you said, but not with sufficient detail for it to make sense to me in terms of how we are restricted.)
> 
> I probably assumed you still know all that stuff from `back in the days™`...

You know, I asked myself the same thing. I remember there being the directory with some native libs, and Felix's loading list for bundles. I may have been actively involved just before the switch to embedded native binaries, or I just encountered the unbundled ones.

cobratbq commented:
I reverted the classloader changes. If you still want to merge because of the pom-change, prly better to squash it :-D ... or drop the PR. I leave it to you.

ibauersachs commented:
> The Java run-time leaves almost no wiggle-room. This also involves `libjitsi` which is shared among products IIUC, so I am not going to go there.

libjitsi is now only used by Jitsi and Jigasi. Both use jitsi-android-osgi and would benefit from simplification. Also, libjitsi just calls `System.loadLibrary`, so as long as this works, it's fine.

> > I probably assumed you still know all that stuff from `back in the days™`...
> 
> You know, I asked myself the same thing. I remember there being the directory with some native libs, and Felix's loading list for bundles.

The list of natives in libs/ were the ones built from within Jitsi's source, and when libjitsi didn't exist yet. The loading list of bundles is something I got rid of since the great repacking since ~2019.

> I may have been actively involved just before the switch to embedded native binaries, or I just encountered the unbundled ones.

The split to libjitsi started the bundling, although still with locally-compiled and committed binaries.

cobratbq commented:
> > The Java run-time leaves almost no wiggle-room. This also involves `libjitsi` which is shared among products IIUC, so I am not going to go there.
> 
> libjitsi is now only used by Jitsi and Jigasi. Both use jitsi-android-osgi and would benefit from simplification. Also, libjitsi just calls `System.loadLibrary`, so as long as this works, it's fine.

FYI, I traced the loading process. `System.loadLibrary` is influenced by `ClassLoader.findLibrary` and even then with some imposed limits. That's how `BundleClassLoader` is able to do what it does. So going back to the original classloader means you'd need to ensure the native binaries are available/locatable on the file-system. Simplification would be easy enough, as long as any users of `libjitsi` (in this particular case) can handle having the native libraries outside the `jar`-file.

Write your own, and you can do all the weird things you want, as long as you can adhere to the restrictions imposed. From my quick look and from memory, the restrictions are roughly:

_After having called `ClassLoader.findLibrary`, you must have the library available and accessible on an absolute filesystem path. (And if `libname` is an absolute path itself, it will immediately load therefore skip `ClassLoader.findLibrary`.)_
